pkg/agent/flowexporter/exporter/exporter_perf_test.go

```diff
@@ -81,7 +83,9 @@ func BenchmarkExportConntrackConns(b *testing.B) {
 	b.ResetTimer()
 	for n := 0; n < b.N; n++ {
 		exp.initFlowExporter()
-		exp.sendFlowRecords()
+		for i := 0; i < int(math.Ceil(testNumOfConns/maxConnsToExport)); i++ {
```

Comment:
@srikartati , could you please help to tell, in order to export all records, is there anything wrong with this change?

The confused thing I observed is, with 20k conns:
benchmark result with pre-allocate slice:
`BenchmarkExportConntrackConns-2  	  100	 10610346 ns/op	23531736 B/op	  8542 allocs/op`
Compare to without pre-allocate slice, just simply append conn to an empty slice by 20k times:
`BenchmarkExportConntrackConns-2   	  100	  4170904 ns/op	 1112738 B/op	  8217 allocs/op`
It's weird that pre-allocate slice does not performs better than we expect.

Pre-allocate slice:
```
(pprof) top
Showing nodes accounting for 2327.28MB, 98.45% of 2364.03MB total
Dropped 96 nodes (cum <= 11.82MB)
Showing top 10 nodes out of 15
      flat  flat%   sum%        cum   cum%
 2218.42MB 93.84% 93.84%  2296.98MB 97.16%  antrea.io/antrea/pkg/agent/flowexporter/exporter.(*flowExporter).sendFlowRecords
   45.05MB  1.91% 95.75%    45.05MB  1.91%  github.com/vmware/go-ipfix/pkg/entities.NewDataRecord
   38.37MB  1.62% 97.37%       57MB  2.41%  antrea.io/antrea/pkg/agent/flowexporter/exporter.addConns
   12.44MB  0.53% 97.90%    12.44MB  0.53%  antrea.io/antrea/pkg/agent/flowexporter/connections.(*connectionStore).AddConnToMap
    7.50MB  0.32% 98.21%    59.55MB  2.52%  antrea.io/antrea/pkg/agent/flowexporter/exporter.(*flowExporter).addConnToSet
       5MB  0.21% 98.42%    77.56MB  3.28%  antrea.io/antrea/pkg/agent/flowexporter/exporter.(*flowExporter).exportConn
    0.50MB 0.021% 98.45%       13MB  0.55%  antrea.io/antrea/pkg/agent/flowexporter/exporter.(*flowExporter).sendDataSet
         0     0% 98.45%  2356.01MB 99.66%  antrea.io/antrea/pkg/agent/flowexporter/exporter.BenchmarkExportConntrackConns
         0     0% 98.45%    58.53MB  2.48%  antrea.io/antrea/pkg/agent/flowexporter/exporter.setupExporter
         0     0% 98.45%       13MB  0.55%  antrea.io/antrea/pkg/ipfix.(*ipfixExportingProcess).SendSet
(pprof) list sendFlowRecord
Total: 2.31GB
ROUTINE ======================== antrea.io/antrea/pkg/agent/flowexporter/exporter.(*flowExporter).sendFlowRecords in /home/vagrant/antrea/pkg/agent/flowexporter/exporter/exporter.go
    2.17GB     2.24GB (flat, cum) 97.16% of Total
         .          .    219:	}
         .          .    220:}
         .          .    221:
         .          .    222:func (exp *flowExporter) sendFlowRecords() (time.Duration, error) {
         .          .    223:	currTime := time.Now()
    2.17GB     2.17GB    224:	expiredConns := make([]flowexporter.Connection, 0, maxConnsToExport*2)
         .          .    225:	var expireTime1, expireTime2 time.Duration
         .        1MB    226:	expiredConns, expireTime1 = exp.conntrackConnStore.GetExpiredConns(expiredConns, currTime, maxConnsToExport)
         .          .    227:	expiredConns, expireTime2 = exp.denyConnStore.GetExpiredConns(expiredConns, currTime, maxConnsToExport)
         .          .    228:	// Select the shorter time out among two connection stores to do the next round of export.
         .          .    229:	nextExpireTime := getMinTime(expireTime1, expireTime2)
         .          .    230:	for i := range expiredConns {
         .    77.56MB    231:		if err := exp.exportConn(&expiredConns[i]); err != nil {
         .          .    232:			klog.ErrorS(err, "Error when sending expired flow record")
         .          .    233:			return nextExpireTime, err
         .          .    234:		}
         .          .    235:	}
         .          .    236:	return nextExpireTime, nil
(pprof) 
```
I'm also confused about how the 2.17GB comes. As a 128 conns slice takes 4.82MB. ceil(20000/64) = 312. 4.82MB * 312 = 1503MB = 1.503GB

Reply:
>I'm also confused about how the 2.17GB comes. As a 128 conns slice takes 4.82MB. ceil(20000/64) = 312. 4.82MB * 312 = 1503MB = 1.503GB

how did you compute 4.82 MB? Did you consider the memory alignment of fields in the struct?

As we are pre-allocating, can we do it once and keep it in `flowExporter` struct to reuse that? Once the flow is exported, it can be safely overwritten. What do you think?
In 20K case, allocated memory in this step exceeds the memory utilized when doing append.

Reply:
> how did you compute 4.82 MB? 

I call `sendFlowRecord` only once. And found the memory usage of the line `expiredConns := make([]flowexporter.Connection, 0, maxConnsToExport*2)` to be 4.82MB.

> As we are pre-allocating, can we do it once and keep it in flowExporter struct to reuse that? Once the flow is exported, it can be safely overwritten. What do you think?

Thanks. That makes a lot of sense to me! The result looks good:
`BenchmarkExportConntrackConns-2   	     100	   3550826 ns/op	  528366 B/op	    8228 allocs/op`
Compare to main:
`BenchmarkExportConntrackConns-2   	     100	   8910469 ns/op	  634736 B/op	    7583 allocs/op`

Do you think it's better to make `exportConns` a field of `flowExporter` or make it a global variable in the pkg? The slice is only used in sendFlowRecord this method.

Reply:
>Do you think it's better to make exportConns a field of flowExporter or make it a global variable in the pkg? The slice is only used in sendFlowRecord this method.

Why do you think the global variable is apt? is it better in terms of memory layout?

Reply:
No, I just don't know what is the best practice. After giving another thought, I think struct field is more appropriate, as we need a different copy of the slice for every `flowExporter`. I have made it a field of `flowExporter` in the latest code.